### ISSUE TITLE
Dart ctor are not inherited by default

### ISSUE THREAD

**Issue by mprobst** [angular/ts2dart#100]
_From @vicb on October 8, 2014 15:36_

In ES6, ctor are inherited by default, see https://github.com/angular/angular/issues/66#issuecomment-58371981

That is if the class inherit from a parent class and this class a no ctor, the parent ctor will be called:

```
class Parent {
  constructor(foo) {
    // ...
  } 
}

class Child extends Parent {
}

// acts as

class Parent {
  constructor(foo) {
    // ...
  } 
}

class Child extends Parent {
   constructor(...args) { super(...args); }
}
```

The equivalent for `constructor(...args) { super(...args); }  has to be generated for Dart

_Copied from original issue: angular/angular#70_

**Comment by mprobst** [comment]
@vicb what would the equivalent Dart code be? super(...args) probably won't work, and cannot be type checked.

**Comment by mprobst** [comment]
_From @vicb on March 26, 2015 7:36_

```js
class Parent {
  constructor(foo: Foo) {
    // ...
  } 
}

class Child extends Parent {
}
```

In this case for Dart you should generate `Child(foo: Foo): super(foo);`

I think you should always have access to the parent class ctor signature in TS ? (which is not the case with Traceur).

**Comment by mprobst** [comment]
Yes, once we can type check the TS code we should have access to signatures.

**Comment by vicb** [comment]
Just got bitten by this:

```
class _UnexpectedValueException extends BaseException {
  constructor(message: string) {
    super(message);
  }
}
```

I first did not declare the ctor.... which was fine with JS

but in Dart it would generate an Error `No constructor '_UnexpectedValueException' with matching arguments declared in class '_UnexpectedValueException'`

**Comment by mprobst** [comment]
Victor: do you think we should copy the parent classes ctor when you're
subclassing? I don't think this would necessarily break something, but it's
a bit complicated and doesn't quite follow the design decisions of Dart.

Victor Berchet <notifications@github.com> schrieb am Fr., 15. Mai 2015 um
12:27 Uhr:

**Comment by vicb** [comment]
@mprobst I see 2 options here:
- throw when a class extends a base class and do not implement have an explicit constructor,
- add a constructor that forwards to `super()` when no constructor is defined - in such a case you would need to copy the signature from the base class which could be either Dart or TS.

**Comment by mprobst** [comment]
@vicb I think throwing an error is too restrictive - Dart does support default constructors with no arguments, if we'd throw an error we'd force users to write a ctor for *every* class that extends a class, which seems overly restrictive.

The other option seems better, or as a third option we could only throw an error for the actual error case (no default zero arguments super ctor). But that'd need complete type information in the transpiler, so this will have to wait a bit.
